Add some basic docs for dummy data in Data Builder (#635)
opensafely/documentation: docs/data-builder-dummy-data.md

# Supplying a dummy dataset

---8<-- 'includes/data-builder-danger-header.md'

Because OpenSAFELY doesn't allow direct access to individual patient
records, researchers must use a **dummy dataset** for developing their analytic

> code on their own computer.

Reviewer: And when pushing to GitHub, no? How about...

```suggestion
code.
```

or alternatively, using the terminology from the [Security](https://docs.opensafely.org/security-levels/) section...

```suggestion
code outside the EHR vendor's secure environment on the OpenSAFELY server.
```

Author: Ah yes, I copied this verbatim from the existing docs, but your suggestions are improvements.

A dummy dataset needs to be similar enough to a real dataset to allow

> the analysis scripts to run successfully, but it must be randomly (or

Reviewer: For consistency, how about...

```suggestion
[the project pipeline](actions-pipelines.md) to run successfully, but it must be randomly (or
```

Reviewer 2: In addition to @iaindillingham's suggestion, I'd probably also emphasise the "not contain data relating to real patients" at the end of this line.

At the least, I'd break this last clause out into its own sentence. It's very important, but is at the end. Maybe even with an admonition like `!!! warning`? [^1]

[^1]: This might reasonably be `!!! danger`. But searching through the Markdown source, `!!! danger` has only been used within the Data Builder docs so far. Therefore there may be similarly-themed text where we're not using `!!! danger`, and for consistency, I'd leave for now, and maybe address that later.

psuedo-randomly) generated and not contain data relating to real
patients.

> In future, Data Builder will be able to generate suitable dummy data

Reviewer 2: question: are we guaranteeing this?

Reviewer 2: suggestion: Perhaps consider splitting this sentence into two (until "but for now") and maybe putting the first part in a `!!! note`.

from a dataset definition, but for now you must generate your own dummy
data file and use the `--dummy-data-file` argument to tell Data Builder
where it is. When running on your own computer (or in the automated

> tests), Data Builder will copy this file to the same place it will put

Reviewer 2: suggestion: For consistency, you might want to refer to this situation in the same way as suggested in [@iaindillingham's comment earlier](https://github.com/opensafely/documentation/pull/635/files#r821938781). That is: referring to "when running outside of a secure EHR vendor environment" or similar.

Reviewer 2: thought: "same place" here is possibly ambiguous. For example, on my computer, I cannot access a vendor's file system. Do you mean something along the lines of "the same relative file path"?

the real data when it is run inside the secure environment. This allows
the rest of the analysis code to run in the same way whether it is
working with real or dummy data.

Data Builder will check that the column names and types of value in the
dummy dataset match what will be in the real dataset and will raise an
error if it finds a discrepancy.

## Example

```yaml title="Minimal Data Builder project YAML example"
version: '3.0'

actions:

  generate_dataset:

>     run: databuilder:v0 generate_dataset --dataset-definition analysis/dataset_definition.py --dummy-data-file dummy_data.csv --output output/dataset.csv

Reviewer: It seems a shame that the most important information is hidden. How about...

```suggestion
    run: >
      databuilder:v0 generate_dataset
        --dataset-definition analysis/dataset_definition.py
        --dummy-data-file dummy_data.csv
        --output output/dataset.csv
```

Author: OH YES! I didn't realise YAML supported this. Thank you! We should be using this everywhere.

Reviewer: [The StackOverflow question about line breaks in YAML](https://stackoverflow.com/questions/3790454/how-do-i-break-a-string-in-yaml-over-multiple-lines) is blub studies at its finest.
